0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+manager nodes. We acknowledge that this approach is not comprehensive,
+as a compromise of a single key manager enclave would reveal past secrets,
+potentially leading to the decryption of the internal state of runtimes
+using them. However, this key derivation method is straightforward and,
+as a result, exceptionally fast.
+
+While master secret rotations and ephemeral secrets aim to rotate the secrets
+to mitigate the impact of a secret compromise, we also want to support
+different types of key derivation, each offering varying levels of security.
+Examples include verifiable secret sharing schemes and dynamic-committee
+proactive secret sharing, where key managers would only hold a share
+of the secret.
+
+This proposal aims to introduce support for the CHUrn-Robust Proactive
+secret sharing scheme (CHURP) and Key Derivation Center (KDC).
+
+## Key manager apps
+
+To facilitate the straightforward addition of new features to the key manager,
+we must first generalize it to support the concurrent execution of multiple
+(independent) applications.
+
+Not all key manager runtimes are required to support all apps, and likewise,
+not all nodes need to run all the apps. In fact, the committee nodes for each
+application should be dynamic, allowing for the addition or removal of nodes
+based on specific requirements.
+
+### Current situation
+
+Currently, the key manager supports two applications: one for generating,
+distributing, and storing master secrets, and the other for ephemeral secrets.
+However, these two applications are not independent, as they both share
+the same key manager policy for secret replication and key derivation.
+
+Issues:
+
+- In the runtime, the logic for key manager status, policy, master secrets,
+  and ephemeral secrets should be decoupled.
+
+- On the host, each application should have its own worker (e.g., a master
+  secret app should have a dedicated worker responsible for participating
+  in the master secret protocol).
+
+### Example apps
+
+Current and future applications:
+
+- Master secrets (for generation and replication master secrets)
+
+- Ephemeral secrets (for generation and replication of ephemeral secrets)
+
+- CPU change (for detecting wether the CPU has changed)
+
+- CHURP (secret sharing scheme)
+
+- Key derivation center (secret sharing scheme)
+
+### App trait
+
+Every application should implement the following trait.
+
+```rust
+/// Key manager application.
+pub trait App {
+  /// Register RPC methods to the dispatcher.
+  ///
+  /// Registered methods are exposed via enclave RPC to the local host
+  /// or remote clients.
+  fn register(&self, ...);
+
+  /// Initialize the application on startup.
+  ///
+  /// Use runtime ID, untrusted local storage, host data, consensus verifier,
+  /// runtime identity, etc. to start the application.
+  fn init(&self, ...);
+}
+```
+
+Each application should register RPC methods, prefixed with the application
```

Comment:
I would specify a naming convention for methods, e.g. `app.Method` to be consistent with other parts that have method names.